File: frontend/src/types/asset.ts
export type AssetType = 'bank' | 'cash' | 'credit_card';

export type Asset = {
  id: number;
  name: string;
  type: AssetType;
  balance: string;
  sort_order: number;
};

export type AssetList = {
  items: Asset[];
  total_balance: string;
};

export type AssetCreateInput = {
  name: string;
  type: AssetType;
  balance: string;
};

export type AssetUpdateInput = {
  name: string;
  type: AssetType;
};

export type AdjustmentTransaction = {
  id: number;
  date: string;
  amount: string;
  entry_kind: 'income' | 'expense';
  entry_type: 'normal' | 'adjustment';
  asset_id: number;
  memo: string | null;
};

export type AssetAdjustResult = {
  asset: Asset;
  transaction: AdjustmentTransaction | null;
};

File: backend/app/schemas/asset.py
from datetime import date as date_type
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class AdjustmentTransactionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    date: date_type
    amount: Decimal
    entry_kind: str
    entry_type: str
    asset_id: int
    memo: str | None


class AssetRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    type: str
    balance: Decimal
    sort_order: int


class AssetListRead(BaseModel):
    items: list[AssetRead]
    total_balance: Decimal


class AssetCreate(BaseModel):
    name: str = Field(min_length=1, max_length=50)
    type: str
    balance: Decimal = Decimal(0)


class AssetUpdate(BaseModel):
    name: str = Field(min_length=1, max_length=50)
    type: str


class AssetAdjust(BaseModel):
    actual_balance: Decimal


class AssetAdjustResult(BaseModel):
    asset: AssetRead
    transaction: AdjustmentTransactionRead | None
